README.md
[![Build Status](https://scrutinizer-ci.com/g/brendt/responsive-images/badges/build.png?b=master)](https://scrutinizer-ci.com/g/brendt/responsive-images/build-status/master) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/brendt/responsive-images/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/brendt/responsive-images/?branch=master)

# Responsive Images

Automatically generate responsive images to work with the `srcset` and `sizes` spec. ([http://responsiveimages.org/](http://responsiveimages.org/))

```sh
composer require brendt/responsive-images
```

## Usage

```php
use Brendt\Image\ResponsiveFactory;

$factory = new ResponsiveFactory();
$image = $factory->create('img/image.jpeg');
```

```html
<img src="<?= $image->src() ?>" 
     srcset="<?= $image->srcset() ?>"/>
```

This sample would generate something like:

```hmtl
<img src="/image.jpeg" 
     srcset="/image-384.jpg 384w,
             /image-768.jpg 768w,
             /image-1152.jpg 1152w,
             /image-1536.jpg 1536w,
             /image.jpg 1920w" />
```

## Configuration

The `ResponsiveFactory` can take a `ResponsiveFactoryConfigurator` object which will set the needed parameters. 
A default configurator `DefaultConfigurator` is provider out of the box, and uses the following parameters:
 
```php
[
    'driver'           => 'gd',
    'includeSource'    => true,
    'publicPath'       => './',
    'sourcePath'       => './',
    'rebase'           => false,
    'enableCache'      => false,
    'optimize'         => false,
    'scaler'           => 'filesize',
    'stepModifier'     => 0.5,
    'minFileSize'      => 5000,
    'maxFileSize'      => null,
    'minWidth'         => 300,
    'maxWidth'         => null,
    'sizes'            => [ 1920, 840, 300 ],
    'optimizerOptions' => [],
]
```

You can override these parameters by providing and array to the `DefaultConfigurator`, 
or create a whole new configurator which implements `ResponsiveFactoryConfigurator`.

```php
$factory = new ResponsiveFactory(new DefaultConfigurator([
    'driver'       => 'imagick',
    'sourcePath'   => './src',
    'publicPath'   => './public',
    'enableCache'  => true,
]));
```

### All configuration options

- `driver`: the image driver to use. Defaults to `gd`. Possible options are `gd` or `imagick`.
- `includeSource`: whether to include the source image in the `srcset`. Defaults to `true`.
- `sourcePath`: the path to load image source files. Defaults to `./`.
- `publicPath`: the path to render image files. Defaults to `./`.
- `rebase`: ignore the path of the requested image when searching in the source directory. Defaults to `false`.
- `enableCache`: enable or disable image caching. Enabling the cache wont' override existing images. Defaults to `false`.
- `optimize`: enable or disable the use of different optimizers (if installed on the system). Defaults to `false`.
- `scaler`: which scaler algorithm to use. Defaults to `filesize`. Possible options are `filesize`, `width` or `sizes`.
- `stepModifier`: a percentage (between 0 and 1) which is used to create different image sizes. The higher this modifier, the more image variations will be rendered. Defaults to `0.5`.
- `minFileSize`: the minimum image filesize in bytes. Defaults to `5000`B (5KB).
- `maxFileSize`: the maximum image filesize in bytes. Defaults to `null`.
- `minWidth`: the minimum image size in pixels. No images with size smaller than this number will be rendered. Defaults to `300` pixels.
- `maxWidth`: the maximum image size in pixels. No images with size smaller than this number will be rendered. Defaults to `null`.
- `sizes`: this parameter is used when the `sizes` scaler is enabled. This scaler will generate a fixed set of sizes, based on this array. 
 The expected values are the widths of the generated images. Defaults to `[]` (empty array). 
- `optimizerOptions`: options to pass to the image optimizer library. See [https://github.com/psliwa/image-optimizer](https://github.com/psliwa/image-optimizer) for more information.

### Paths

The `sourcePath` parameter is used to define where image source files are located. 
In case of the first example and above configuration, the image file should be saved in `./src/img/image.jpeg`.

The `publicPath` parameter is used to save rendered images into. This path should be the public directory of your website.
The above example would render images into `./public/img/image.jpeg`. 

#### Path rebasing

When the `rebase` option is enabled, source file lookup will differ: only the filename is used to search the file in the 
 source directory. An example would be the following.
 
```php
// Without rebase

$options = [
    'sourcePath' => './src/images',
    'publicPath' => './public',
];

$image = $factory->create('/img/responsive/image.jpeg');

// Source file is searched in './src/images/img/responsive/image.jpeg' 
// Public files are saved in './public/img/responsive/image-x.jpg'
``` 

```php
// With rebase

$options = [
    'sourcePath' => './src/images',
    'publicPath' => './public',
    'rebase'     => true,
];

$image = $factory->create('/img/responsive/image.jpeg');

// Source file is searched in './src/images/image.jpeg'  
// Public files are saved in './public/img/responsive/image-x.jpg'
```


CHANGELOG.md
# Changelog

## 1.9.1

- Support PHP 7.0

## 1.9.0

- Refactor code directories.

## 1.8.2

- Cleanup Responsive Factory code.
- Add caching test.

## 1.8.1

- Bugfix for extension of cached files not correctly loading.

## 1.8.0

- Add `includeSource` parameter.

## 1.7.3

- Bugfix for invalid cache path when rebase was enabled.

## 1.7.2

- Bugfix for wrong composer sources.

## 1.7.1

- Bugfix for srcset not correctly rendering.

## 1.7.0

- Add `sizes` scaler. This scaler can be configured to generate a fixed set of images based on pre-defined widths, 
 provided via the `sizes` parameter. See the README for more information.

## 1.6.0

- Add `rebase` parameter.
- Add `maxFileSize` parameter.
- Add `maxWidth` parameter.

## 1.5.1

- Fix a bug so that unkown file exceptions are correctly thrown.

## 1.5.0

- Removed AMP dependencies.

## 1.4.1

- Temporary remove Amp because of BC breaking changes with each update.

## 1.4.0

- Add optimizer options support

## 1.3.2

- Add fixed amphp versions because of BC breaking changes in their library.

## 1.3.1

- Several bug fixes and optimizations when running enabling the `async` option.

## 1.3.0

- Add `async` option to downscale images in separate processes. PHP's `pcntl` extension is required for this to work.

## 1.2.2

- Add simple construct for the ResponsiveFactory.

## 1.2.1

- Fixed bug with the cache not being searched in the public path.

## 1.2.0

- Update minimum requirements to PHP 7.1.
- Refactor scalers to only calculate file sizes and not handle image saving.
- Add `optimize` parameter to run image optimizers. Defaults to `false`.

## 1.1.2

- Improve caching, scalers are now not used when images are already cached.
- Set default `minFileSize` to 5 KB.

## 1.1.0

- Add Scaler support
- Add FileSize scaler as default.

To use the old scaler:

```php
$factory = new ResponsiveFactory(new DefaultConfigurator([
    'scaler' => 'width',
]));
```

The new Scaler is called `filesize` and set by default, this scaler differs from the width scaler because it scales down 
 based on file size instead of the width of an image.

```php
$factory = new ResponsiveFactory(new DefaultConfigurator([
    'scaler' => 'filesize',
]));
```

**Note**: the `stepModifier` parameter is now used different, the higher this modifier, the more images will be generated.

## 1.0.0

- Use PSR naming standards.


src/Config/DefaultConfigurator.php
<?php

namespace Brendt\Image\Config;

use Brendt\Image\Exception\InvalidConfigurationException;
use Brendt\Image\ResponsiveFactory;
use Brendt\Image\Scaler\AbstractScaler;
use Brendt\Image\Scaler\FileSizeScaler;
use Brendt\Image\Scaler\Scaler;
use Brendt\Image\Scaler\SizesScaler;
use Brendt\Image\Scaler\WidthScaler;

class DefaultConfigurator implements ResponsiveFactoryConfigurator
{

    /**
     * The default config
     *
     * @var array
     */
    protected $config = [
        'driver'           => 'gd',
        'publicPath'       => './',
        'sourcePath'       => './',
        'rebase'           => false,
        'enableCache'      => false,
        'optimize'         => false,
        'scaler'           => 'filesize',
        'stepModifier'     => 0.5,
        'minFileSize'      => 5000,
        'maxFileSize'      => null,
        'minWidth'         => 300,
        'maxWidth'         => null,
        'sizes'            => [],
        'optimizerOptions' => [],
        'includeSource'    => true,
    ];

    /**
     * ResponsiveFactoryConfigurator constructor.
     *
     * @param array $config
     *
     * @throws InvalidConfigurationException
     */
    public function __construct(array $config = []) {
        if (isset($config['driver']) && !in_array($config['driver'], ['gd', 'imagick'])) {
            throw new InvalidConfigurationException('Invalid driver. Possible drivers are `gd` and `imagick`');
        }

        $this->config = array_merge($this->config, $config);
    }

    /**
     * @param ResponsiveFactory $factory
     *
     * @return void
     */
    public function configure(ResponsiveFactory $factory) {
        /** @var AbstractScaler $scaler */
        switch ($this->config['scaler']) {
            case 'filesize':
                $scaler = new FileSizeScaler($this);
                break;
            case 'width':
                $scaler = new WidthScaler($this);
                break;
            case 'sizes':
            default:
                $scaler = new SizesScaler($this);
                $scaler->setSizes($this->config['sizes']);
                break;
        }

        $factory
            ->setDriver($this->config['driver'])
            ->setPublicPath($this->config['publicPath'])
            ->setSourcePath($this->config['sourcePath'])
            ->setRebase($this->config['rebase'])
            ->setEnableCache($this->config['enableCache'])
            ->setOptimize($this->config['optimize'])
            ->setOptimizerOptions($this->config['optimizerOptions'])
            ->setScaler($scaler);
    }

    /**
     * @param Scaler $scaler
     *
     * @return Scaler
     */
    public function configureScaler(Scaler $scaler) {
        $scaler
            ->setIncludeSource($this->config['includeSource'])
            ->setMinFileSize($this->config['minFileSize'])
            ->setMinWidth($this->config['minWidth'])
            ->setMaxFileSize($this->config['maxFileSize'])
            ->setMaxWidth($this->config['maxWidth'])
            ->setStepModifier($this->config['stepModifier']);

        return $scaler;
    }

    /**
     * @return array
     */
    public function getConfig() {
        return $this->config;
    }

    /**
     * @param $key
     *
     * @return mixed|null
     */
    public function get($key) {
        return isset($this->config[$key]) ? $this->config[$key] : null;
    }
}
